### PR TITLE
Modularize Hive Metastore

### PR DESCRIPTION
**Pull request by islamismailov**
```
== RELEASE NOTES ==

Hive Changes
* Metastore interface is separated into a separate module to reduce monolithicness 
```

### CODE REVIEW

**Comment by islamismailov** [comment]
I can move ExtendedHiveMetastore into the new module too, and possibly the SemiTransactionalHiveMetastore too. I am afraid that might be too coupled and monolithic (e.g. classes like HiveType etc. I will try to do in separate commits so we can reason about it in pieces, as it might be easier.

**Comment by islamismailov** [comment]
Here are the three commits:

1. Modularize only necessary components
2. Add ExtendedHiveMetastore pieces to the module
3. Add SemiTransactionalHiveMetastore pieces to the module

File: pom.xml [hunk 1]
```diff
@@ -112,6 +112,7 @@
         <module>presto-client</module>
         <module>presto-parser</module>
         <module>presto-main</module>
+        <module>presto-metastore</module>
```
Comment (highker) [inline pom.xml]: How about calling it `presto-hive-metastore`?

**Comment by islamismailov** [comment]
This should be ready for the review now :)

File: pom.xml [hunk 2]
```diff
@@ -375,6 +376,19 @@
                 <type>test-jar</type>
             </dependency>
 
+            <dependency>
+                <groupId>com.facebook.presto</groupId>
+                <artifactId>presto-hive-metastore</artifactId>
+                <version>${project.version}</version>
+            </dependency>
+
+            <dependency>
```
Comment (highker) [inline pom.xml]: This doesn't seem to be necessary

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java [hunk 3]
```diff
@@ -173,4 +167,9 @@ public String toString()
                     .toString();
         }
     }
+
+    static {
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java]: Can we keep this as it was?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java [hunk 4]
```diff
@@ -33,23 +33,17 @@
...
-
     private final HdfsConfiguration hdfsConfiguration;
     private final HdfsAuthentication hdfsAuthentication;
     private final boolean verifyChecksum;
-
     @Inject
     public HdfsEnvironment(
             HdfsConfiguration hdfsConfiguration,
-            HiveClientConfig config,
+            MetastoreClientConfig metastoreClientConfig,
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java]: just call this `config`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java [hunk 5]
```diff
@@ -60,6 +50,16 @@ public HiveBasicStatistics(
         this.onDiskDataSizeInBytes = requireNonNull(onDiskDataSizeInBytes, "onDiskDataSizeInBytes is null");
     }
 
+    public static HiveBasicStatistics createEmptyStatistics()
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java]: keep as is

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveColumnHandle.java [hunk 6]
```diff
@@ -47,27 +47,16 @@
     public static final String BUCKET_COLUMN_NAME = "$bucket";
     public static final HiveType BUCKET_HIVE_TYPE = HIVE_INT;
     public static final TypeSignature BUCKET_TYPE_SIGNATURE = BUCKET_HIVE_TYPE.getTypeSignature();
-
-    private static final String UPDATE_ROW_ID_COLUMN_NAME = "$shard_row_id";
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveColumnHandle.java]: Keep this file as is

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreErrorCode.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    HIVE_INVALID_METADATA(12, EXTERNAL),
+    HIVE_FILESYSTEM_ERROR(16, EXTERNAL),
+    HIVE_UNSUPPORTED_FORMAT(19, EXTERNAL),
+    /**/;
+
+    private final ErrorCode errorCode;
+
+    MetastoreErrorCode(int code, ErrorType type)
+    {
+        errorCode = new ErrorCode(code + 0x0100_0000, name(), type);
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreErrorCode.java]: Shall we make `0x0100_0000` a constant and use it in both `MetastoreErrorCode` and `HiveErrorCode`?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreErrorCode.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+public enum MetastoreErrorCode
+        implements ErrorCodeSupplier
+{
+    HIVE_METASTORE_ERROR(0, EXTERNAL),
+    HIVE_TABLE_OFFLINE(2, USER_ERROR),
+    HIVE_PARTITION_OFFLINE(6, USER_ERROR),
+    HIVE_INVALID_METADATA(12, EXTERNAL),
+    HIVE_FILESYSTEM_ERROR(16, EXTERNAL),
+    HIVE_UNSUPPORTED_FORMAT(19, EXTERNAL),
+    /**/;
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreErrorCode.java]: Let's complete the comment to say "Shared error code with HiveErrorCode".

File: presto-hive/pom.xml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (highker) [inline presto-hive/pom.xml]: accident change?

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveFileSystem.java [hunk 10]
```diff
@@ -113,9 +113,8 @@
 
 public abstract class AbstractTestHiveFileSystem
 {
-    private static final HdfsContext TESTING_CONTEXT = new HdfsContext(new ConnectorIdentity("test", Optional.empty(), Optional.empty()));
     public static final SplitSchedulingContext SPLIT_SCHEDULING_CONTEXT = new SplitSchedulingContext(UNGROUPED_SCHEDULING, false);
-
+    private static final HdfsContext TESTING_CONTEXT = new HdfsContext(new ConnectorIdentity("test", Optional.empty(), Optional.empty()));
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveFileSystem.java]: unrelated change?

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 11]
```diff
@@ -58,18 +58,16 @@
...
-    }
-
     public static final String HIVE_CATALOG = "hive";
     public static final String HIVE_BUCKETED_CATALOG = "hive_bucketed";
     public static final String TPCH_SCHEMA = "tpch";
     public static final String TPCH_BUCKETED_SCHEMA = "tpch_bucketed";
+    private static final Logger log = Logger.get(HiveQueryRunner.class);
     private static final String TEMPORARY_TABLE_SCHEMA = "__temporary_tables__";
     private static final DateTimeZone TIME_ZONE = DateTimeZone.forID("America/Bahia_Banderas");
+    private HiveQueryRunner()
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: unrelated change?

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 12]
```diff
@@ -58,18 +58,16 @@
...
-
-    private HiveQueryRunner()
-    {
-    }
-
     public static final String HIVE_CATALOG = "hive";
     public static final String HIVE_BUCKETED_CATALOG = "hive_bucketed";
     public static final String TPCH_SCHEMA = "tpch";
     public static final String TPCH_BUCKETED_SCHEMA = "tpch_bucketed";
+    private static final Logger log = Logger.get(HiveQueryRunner.class);
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: unrelated change?

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveTestUtils.java [hunk 13]
```diff
@@ -68,20 +68,13 @@
 
 public final class HiveTestUtils
 {
-    private HiveTestUtils()
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveTestUtils.java]: Let's make sure this file has no unrelated code movement

File: presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestHiveMetastoreUtil.java [hunk 14]
```diff
@@ -70,11 +71,6 @@
             "view original text",
             "view extended text",
             "MANAGED_TABLE");
-
-    static {
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestHiveMetastoreUtil.java]: same

File: presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestThriftHiveMetastoreUtil.java [hunk 15]
```diff
@@ -59,8 +59,13 @@
 import static org.apache.hadoop.hive.serde.serdeConstants.STRING_TYPE_NAME;
 import static org.testng.Assert.assertEquals;
 
-public class TestThriftMetastoreUtil
+public class TestThriftHiveMetastoreUtil
 {
+    private static void testBasicStatisticsRoundTrip(HiveBasicStatistics expected)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestThriftHiveMetastoreUtil.java]: same

**Comment by highker** [comment]
@islamismailov, could you address the comments?

**Comment by islamismailov** [comment]
oh whoops! I thought i did `push` - sorry i forgot to press enter!

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 16]
```diff
@@ -162,10 +161,9 @@
 
 public final class HiveUtil
 {
-    private static final Pattern DEFAULT_HIVE_COLUMN_NAME_PATTERN = Pattern.compile("_col\\d+");
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Keep as is

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveFileSystem.java [hunk 17]
```diff
@@ -115,7 +115,6 @@
 {
     private static final HdfsContext TESTING_CONTEXT = new HdfsContext(new ConnectorIdentity("test", Optional.empty(), Optional.empty()));
     public static final SplitSchedulingContext SPLIT_SCHEDULING_CONTEXT = new SplitSchedulingContext(UNGROUPED_SCHEDULING, false);
-
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveFileSystem.java]: keep this line

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveTestUtils.java [hunk 18]
```diff
@@ -119,15 +120,14 @@ public String formatRowExpression(ConnectorSession session, RowExpression expres
     }
 
     public static final HiveClientConfig HIVE_CLIENT_CONFIG = new HiveClientConfig();
-
-    public static final HdfsEnvironment HDFS_ENVIRONMENT = createTestHdfsEnvironment(HIVE_CLIENT_CONFIG);
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveTestUtils.java]: Do not move this around

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePartitionManager.java [hunk 19]
```diff
@@ -69,9 +69,8 @@
             Optional.empty());
 
     private static final List<String> PARTITIONS = ImmutableList.of("ds=2019-07-23", "ds=2019-08-23");
-
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePartitionManager.java]: keep this line

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePartitionManager.java [hunk 20]
```diff
@@ -69,9 +69,8 @@
...
-    private HivePartitionManager hivePartitionManager = new HivePartitionManager(new TestingTypeManager(), new HiveClientConfig());
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePartitionManager.java]: keep as is

**Comment by islamismailov** [comment]
rebasing...

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java [hunk 21]
```diff
@@ -41,11 +41,10 @@
     private final HdfsConfiguration hdfsConfiguration;
     private final HdfsAuthentication hdfsAuthentication;
     private final boolean verifyChecksum;
-
```
Comment (highker) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HdfsEnvironment.java]: keep this line
